docs/device-size-change-detection.md

```diff
@@ -0,0 +1,32 @@
+# Blockdevice size change detection
+
+## Introduction
+
+This document presents the design for detecting changes in the capacity (or size) of
+blocdevices in NDM.
+
+## Summary
+
+When a new blockdevice is detected, NDM collects all its details using different probes. One of the parameter collected is the size of the blockdevice. This is done using the **sysfs probe**. While the size of the device is detected when its first discovered, NDM remains oblivious to any changes that might happen to the size of the device once it has been scanned and added. Chaging the size of the device is quite common and easy to perfrom in cloud environments. Having out-dated info about the device size causes provisioning issues.Currently, the only way to update the blockdevice with the new size is to restart NDM.
+
+This documents presents a system to detect changes to device capacity and update the upstream BlockDevice CR in etcd accordingly.
+
+### Goals
+
+- Detect changes to blockdevice capacity
+- Retrieve new size and update the BlockDevice CR in etcd
+
+## Design
+
+The capacity of a blockdevice is currently fetched by the *sysfs* probe by reading the `size` attribute
+of the blockdevice which is available as a file. This is only done when NDM first scans for devices or when it detects a new device has been added. NDM uses the *udev* system provided by the kernel events to capture events of devices being added or removed and act accordingly. The kernel uses the *udev* system to  inform the user-space programs about various events. These events include addition or removal of a blockdevice as well as events about any changes to a blockdevice. Whenever the size of a blockdevice changes, a change event is emitted by the kernel. Thus, size change detection is done by capturing change events from the *udev* system. However, change events are emitted for multiple reasons and not only a change in the device capacity. It has also been found that a significant amount of change events are emitted on VMs on cloud. This means that there's a large amount of noise that needs to be filtered out. While it is possible to filter out change events based on the subsystem (to recieve events from the blockdevice subsystem only), it is not possible to filter them futher to receive events only for blockdevice capacity change.
+
+To capture change events from *udev*, the *sysfs* probe subscribes to change events using the existing *udevevent* package. On receiving a udev event, the sysfs probe processes the event to get the list of devices that were modified and creates a corresponding `controller.EventMessage` with the `requestedProbes` field containing only *sysfs* probe since only the size needs to be updated. This message is then pushed into `controller.EventMessageChannel` and further processed by the event handler.
```

Comment:
Are we planning to put any filtering mechanism to check if the device/size was changed,? and then send the event message.

Reply:
Also, can we have multiple subscribers to the udev event package? 

If two probes subscribe to udev event we need to make sure, if at some point, udev probe also decides to subscribe, it should exclude the sysfs probe, other wise it will result in same probe being run twice for the same data.

Reply:
yes, it is possible to have multiple subscribers to the udev events. But the subscription can also filter out events based on the type (add, remove, change). I planned to let the udevprobe only listen to add/remove events and the sysfs probe only to change events. This would prevent any overlap.

Reply:
sure. that sounds good.